From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Maks Orlovich <morlovich@chromium.org>
Date: Thu, 22 Oct 2026 18:05:58 -0700
Subject: HttpCache::Transaction reject overlong range responses

...that is, those that don't match the length of their own
Content-Range.

Bug: 572055997
Change-Id: I7a0ca8652462fa890cf04944c21e0674d3d21815
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8608994
Reviewed-by: Kenichi Ishibashi <bashi@chromium.org>
Commit-Queue: Maks Orlovich <morlovich@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1725388}

diff --git a/net/http/http_cache_transaction.cc b/net/http/http_cache_transaction.cc
index d8538596b007ba43fdb17e27e4d0221db53b8262..94d460fc777f2449e453e508963da2aab11cc9e6 100644
--- a/net/http/http_cache_transaction.cc
+++ b/net/http/http_cache_transaction.cc
@@ -2547,6 +2547,10 @@ int HttpCache::Transaction::DoNetworkReadCacheWriteComplete(int result) {
 int HttpCache::Transaction::DoPartialNetworkReadCompleted(int result) {
   DCHECK(partial_);
 
+  if (result > 0 && handling_206_ && !partial_->IsNetworkReadOK(result)) {
+    result = ERR_CONTENT_LENGTH_MISMATCH;
+  }
+
   // Go to the next range if nothing returned or return the result.
   // TODO(shivanisha) Simplify this condition if possible. It was introduced
   // in https://codereview.chromium.org/545101
diff --git a/net/http/http_cache_unittest.cc b/net/http/http_cache_unittest.cc
index 2f3a1d86c7c768b14c235274c2b4e74362b600a3..64490780f886a7326f7f03ba78549a087f2bc8a0 100644
--- a/net/http/http_cache_unittest.cc
+++ b/net/http/http_cache_unittest.cc
@@ -2476,6 +2476,89 @@ TEST_F(HttpCacheRangeGetTest, FullAfterPartial) {
   }
 }
 
+// Range physical length not matching what's specified in Content-Range.
+// See crbug.com/572055997
+TEST_F(HttpCacheRangeGetTest, RangeOverrun) {
+  for (const bool use_cache : {true, false}) {
+    MockHttpCache cache;
+    SCOPED_TRACE(use_cache);
+
+    ScopedMockTransaction transaction(kRangeGET_TransactionOK);
+    transaction.request_headers = "Range: bytes = 0-9\r\n" EXTRA_HEADER;
+    std::string response_headers = transaction.response_headers;
+    if (!use_cache) {
+      response_headers += "Cache-Control: no-store\n";
+    }
+    transaction.response_headers = response_headers.c_str();
+
+    transaction.handler = base::BindRepeating(
+        [](const HttpRequestInfo* request, std::string* response_status,
+           std::string* response_headers, std::string* response_data) -> void {
+          RangeTransactionServer::RangeHandler(request, response_status,
+                                               response_headers, response_data);
+          *response_data += "extra stuff";
+        });
+
+    {
+      MockHttpRequest request(transaction);
+      std::unique_ptr<HttpTransaction> http_trans = cache.CreateTransaction();
+      TestCompletionCallback callback;
+      int rv = callback.GetResult(
+          http_trans->Start(&request, callback.callback(), NetLogWithSource()));
+      ASSERT_EQ(rv, OK);
+
+      TestCompletionCallback read_callback;
+      auto buf = base::MakeRefCounted<IOBufferWithSize>(256);
+      rv = read_callback.GetResult(
+          http_trans->Read(buf.get(), 256, read_callback.callback()));
+      EXPECT_EQ(rv, ERR_CONTENT_LENGTH_MISMATCH);
+    }
+
+    {
+      // Test it over two reads as well.
+      MockHttpRequest request(transaction);
+      std::unique_ptr<HttpTransaction> http_trans = cache.CreateTransaction();
+      TestCompletionCallback callback;
+      int rv = callback.GetResult(
+          http_trans->Start(&request, callback.callback(), NetLogWithSource()));
+      ASSERT_EQ(rv, OK);
+
+      TestCompletionCallback read_callback;
+      auto buf = base::MakeRefCounted<IOBufferWithSize>(10);
+      rv = read_callback.GetResult(
+          http_trans->Read(buf.get(), 10, read_callback.callback()));
+      ASSERT_EQ(rv, 10);
+      EXPECT_EQ(base::as_string_view(buf->first(10)),
+                std::string_view("rg: 00-09 "));
+
+      TestCompletionCallback read_callback2;
+      rv = read_callback2.GetResult(
+          http_trans->Read(buf.get(), 10, read_callback2.callback()));
+      EXPECT_EQ(rv, ERR_CONTENT_LENGTH_MISMATCH);
+    }
+
+    {
+      // Try again, without any weirdness.
+      transaction.handler =
+          base::BindRepeating(&RangeTransactionServer::RangeHandler);
+      MockHttpRequest request(transaction);
+      std::unique_ptr<HttpTransaction> http_trans = cache.CreateTransaction();
+      TestCompletionCallback callback;
+      int rv = callback.GetResult(
+          http_trans->Start(&request, callback.callback(), NetLogWithSource()));
+      ASSERT_EQ(rv, OK);
+
+      TestCompletionCallback read_callback;
+      auto buf = base::MakeRefCounted<IOBufferWithSize>(256);
+      rv = read_callback.GetResult(
+          http_trans->Read(buf.get(), 256, read_callback.callback()));
+      ASSERT_EQ(rv, 10);
+      EXPECT_EQ(base::as_string_view(buf->first(10)),
+                std::string_view("rg: 00-09 "));
+    }
+  }
+}
+
 // Tests that when a range request transaction becomes a writer for the first
 // range and then fails conditionalization for the next range and decides to
 // doom the entry, then there should not be a dcheck assertion hit.
diff --git a/net/http/http_cache_writers.cc b/net/http/http_cache_writers.cc
index d4c931aa6dcf9eca2f11eddf544f04bf0663c7d1..41ed3dc57650076d113891cd5d6c8efcd27d01e2 100644
--- a/net/http/http_cache_writers.cc
+++ b/net/http/http_cache_writers.cc
@@ -371,6 +371,18 @@ HttpCache::Writers::WaitingForRead::WaitingForRead(
 HttpCache::Writers::WaitingForRead::~WaitingForRead() = default;
 HttpCache::Writers::WaitingForRead::WaitingForRead(WaitingForRead&&) = default;
 
+PartialData* HttpCache::Writers::GetPartial() const {
+  // The active transaction must be alive if this is a partial request, as
+  // partial requests are exclusive and hence will always be the active
+  // transaction.
+  // TODO(shivanisha): When partial requests support parallel writing, this
+  // assumption will not be true.
+  if (active_transaction_) {
+    return all_writers_.find(active_transaction_)->second.partial;
+  }
+  return nullptr;
+}
+
 int HttpCache::Writers::DoLoop(int result) {
   DCHECK_NE(State::UNSET, next_state_);
   DCHECK_NE(State::NONE, next_state_);
@@ -441,6 +453,11 @@ int HttpCache::Writers::DoNetworkRead() {
 }
 
 int HttpCache::Writers::DoNetworkReadComplete(int result) {
+  PartialData* partial = GetPartial();
+  if (result > 0 && partial && !partial->IsNetworkReadOK(result)) {
+    result = ERR_CONTENT_LENGTH_MISMATCH;
+  }
+
   if (result < 0) {
     next_state_ = State::NONE;
     OnNetworkReadFailure(result);
@@ -479,15 +496,7 @@ int HttpCache::Writers::DoCacheWriteData(int num_bytes) {
 
   int rv = 0;
 
-  PartialData* partial = nullptr;
-  // The active transaction must be alive if this is a partial request, as
-  // partial requests are exclusive and hence will always be the active
-  // transaction.
-  // TODO(shivanisha): When partial requests support parallel writing, this
-  // assumption will not be true.
-  if (active_transaction_) {
-    partial = all_writers_.find(active_transaction_)->second.partial;
-  }
+  PartialData* partial = GetPartial();
 
   if (!partial) {
     last_disk_cache_access_start_time_ = base::TimeTicks::Now();
diff --git a/net/http/http_cache_writers.h b/net/http/http_cache_writers.h
index 6879147ad504a6a8b58ed6c3edfb8204db97f7f5..16641334dcf176d2e46ba7280bea5a7c66916315 100644
--- a/net/http/http_cache_writers.h
+++ b/net/http/http_cache_writers.h
@@ -184,6 +184,8 @@ class NET_EXPORT_PRIVATE HttpCache::Writers {
 
   using TransactionMap = std::map<Transaction*, TransactionInfo>;
 
+  PartialData* GetPartial() const;
+
   // Runs the state transition loop. Resets and calls |callback_| on exit,
   // unless the return value is ERR_IO_PENDING.
   int DoLoop(int result);
diff --git a/net/http/partial_data.cc b/net/http/partial_data.cc
index ab82ed9a98793dd954d8a6d369b508dd4b471fb7..7b09f033ead972e228625a9bda71ed2a7d2e2e7a 100644
--- a/net/http/partial_data.cc
+++ b/net/http/partial_data.cc
@@ -430,11 +430,26 @@ int PartialData::CacheRead(disk_cache::Entry* entry,
   return rv;
 }
 
+bool PartialData::IsNetworkReadOK(int data_len) const {
+  // Check to see if the network has overrun the range.
+  if (data_len) {
+    base::CheckedNumeric<int64_t> write_end_inclusive = current_range_start_;
+    write_end_inclusive += (data_len - 1);
+    int64_t write_end_inclusive_extracted;
+    if (!write_end_inclusive.AssignIfValid(&write_end_inclusive_extracted) ||
+        write_end_inclusive_extracted > current_range_end_) {
+      return false;
+    }
+  }
+  return true;
+}
+
 int PartialData::CacheWrite(disk_cache::Entry* entry,
                             IOBuffer* data,
                             int data_len,
                             CompletionOnceCallback callback) {
   DVLOG(3) << "To write: " << data_len;
+
   if (sparse_entry_) {
     return entry->WriteSparseData(current_range_start_, data, data_len,
                                   std::move(callback));
diff --git a/net/http/partial_data.h b/net/http/partial_data.h
index 4094fcf9c5c31d5040a48b86a364181fcbf98c25..d220e569e015993cada8400ef224887a20335647 100644
--- a/net/http/partial_data.h
+++ b/net/http/partial_data.h
@@ -109,6 +109,11 @@ class PartialData {
                 int data_len,
                 CompletionOnceCallback callback);
 
+  // Returns true if the returned length from network is reasonable.
+  // Expected to be called before deciding to transition towards cache
+  // writing.
+  bool IsNetworkReadOK(int data_len) const;
+
   // Writes |data_len| bytes to cache. This is basically a wrapper around the
   // API of the cache that provides the right arguments for the current range.
   int CacheWrite(disk_cache::Entry* entry,
@@ -137,7 +142,7 @@ class PartialData {
 
   // The portion we're trying to get, either from cache or network.
   int64_t current_range_start_ = 0;
-  int64_t current_range_end_ = 0;
+  int64_t current_range_end_ = 0;  // Inclusive.
 
   // Next portion available in the cache --- this may be what's currently being
   // read, or the next thing that will be read if the current network portion
